### DISCUSSION TITLE
How to filter realtime rows

### DISCUSSION THREAD

**Discussion by kachar** [tmm/react-supabase#17] [Q&A]
Is there a way to filter the records and show only a fraction of them using `useRealtime` hook

Scenario:

I've got tables `groups` and `items` where each item belongs to a single group

```sql
create table public.groups (
  id uuid primary key default uuid_generate_v4(),
  name text
);

create table public.items (
  id uuid primary key default uuid_generate_v4(),
  group_id uuid not null references public.groups on delete cascade,
  name text
);

create index item_group_id_idx on items (group_id);

alter publication supabase_realtime add table items;
```

Each group gets their own url and want to show previous records and update as new records are inserted in the `items` table in the given group.

The current behavior is that I get **all** `items` from **all** `groups` in the response

```tsx
const [{ data, error, fetching }] = useRealtime<Item>(
  'items',
  (data, payload) => data.id === payload.id,
)

```

**Comment by tmm** [comment]
Correct, currently you cannot filter. Could add a `filter` option that is passed through to the internal `useSelect`.

The challenge with that is it will filter the initially fetched data, but not the real-time updates since Supabase doesn't provide a method for filtering the events.

Right now, I recommend filtering client-side after data is fetched or compose your own hook with `useSelect` and `useSubscription`.

**Reply by kachar** [reply]
Yeah, it will be mostly used for the initial fetching and rendering of the page. Now it fetches all data at all times.

For the realtime subscription we can filter in the `compareFn`

```tsx
const groupId = '...'
const [{ data, error, fetching }] = useRealtime<Item>(
    'items',
    (data, payload) => data.id === payload.id && data.group_id === groupId,
  )
```

Thanks for the suggestions.
If you think the `filter` option will be useful addition to the library I can prepare a PR

**Reply by tmm** [reply]
Go for it. I think that could be useful.

**Reply by kachar** [reply]
@tmm I've got a question and I better ask before submitting the PR

We've got the option to add the config as 3rd param and be backwards compatible or introduce a small breaking change in terms of better code readability and introduction of default values.

```tsx
export type UseRealtimeConfig<Data = any> = {
    select?: UseSelectConfig<Data>
    subscription?: UseSubscriptionConfig
}

const [{ data, error, fetching }] =  useRealtime('items', (a, b) => a.id === b.id, { select, subscription })

// hard to skip the compareFn arg and use it's defaults, but pass a config arg
const [{ data, error, fetching }] =  useRealtime('items', undefined, { select, subscription })
```

or we can wrap the `compareFn` inside the `config` option and end up with better definition of the hook params

```tsx
export type UseRealtimeConfig<Data = any> = {
    select?: UseSelectConfig<Data>
    subscription?: UseSubscriptionConfig
    compareFn?: UseRealtimeCompareFn<Data>
}

const [{ data, error, fetching }] =  useRealtime('items', { 
    select,
    subscription,
    compareFn: (a, b) => a.id === b.id,
})

// or skip overriding compareFn and use it's defaults
const [{ data, error, fetching }] =  useRealtime('items', { 
    select: { 
        filter: (query) => query.eq('status', 'completed'),
    },
    subscription: { event: 'INSERT' }
})
```

**Reply by kachar** [reply]
Actually the `compareFn` is only used in the `subscription` reducer action type and never in `fetch`, which makes it more or less part of the subscription defaults.

```tsx
export type UseRealtimeConfig<Data = any> = {
    select?: UseSelectConfig<Data>
    subscription?: UseSubscriptionConfig & { compareFn?: UseRealtimeCompareFn<Data> }
}

const [{ data, error, fetching }] =  useRealtime('items', { 
    select: { 
        filter: (query) => query.eq('status', 'completed'),
    },
    subscription: {
        event: 'INSERT', 
        compareFn: (a, b) => a.id === b.id,
    }
})
```

**Reply by tmm** [reply]
This looks good, but what happens if you want to add another compare function for another event type?

Let's think about this more. I think you are onto something though.

I'm okay releasing a breaking change at this point if it improves the API because the library isn't super popular yet.

**Reply by kachar** [reply]
> what happens if you want to add another compare function for another event type?

The user may have multiple `useRealtime` hooks for different events within the same component.

```tsx
const [{ dataInsert }] =  useRealtime('items', {
    subscription: {
        event: 'INSERT', 
        compareFn: (a, b) => a.id === b.id,
    }
})
const [{ dataDelete }] =  useRealtime('items', {
    subscription: {
        event: 'DELETE', 
        compareFn: (a, b) => a.deleted === b.deleted,
    }
})
```

**Reply by tmm** [reply]
Got it. It seems like this is starting to look similar to [`useSubscription`](https://react-supabase.vercel.app/documentation/realtime/use-subscription) in its current form:

```tsx
useSubscription(
  (payload) => {
    console.log('Change received!', payload)
  },
  {
    event: 'INSERT',
    table: 'items',
  },
)
```

Another approach could be making the compare function either a function:

```tsx
const [result, reexecute] = useRealtime(
  'items',
  (data, payload) => data.username === payload.username,
)
```

Or object:

```tsx
const [result, reexecute] = useRealtime(
  'items',
  {
    delete: (data, payload) => data.deleted === payload.deleted,
    insert: (data, payload) => data.id === payload.id,
    update: (data, payload) => data.username === payload.username,
  },
)
```

What do you think about this approach?

**Reply by kachar** [reply]
@tmm After taking some time off the problem I came up with a more elegant solution by providing either the table name or `useSelect` options with a table name as first argument of `useRealtime`

You might find the PR at https://github.com/tmm/react-supabase/pull/21

```tsx highlight=6
const [result, reexecute] = useRealtime('todos')

// or

const [result, reexecute] = useRealtime(
  {
    table: 'todos',
    columns: 'id, username, description',
    filter: (query) => query.eq('completed', false),
  },
  (data, payload) => data.username === payload.username,
)
```

In this way we don't break backward compatability and still support the quick way of defining the table argument
